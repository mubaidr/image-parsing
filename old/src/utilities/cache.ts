import Cache from '../utilities/@classes/Cache'

const cache = new Cache()

export { cache }
