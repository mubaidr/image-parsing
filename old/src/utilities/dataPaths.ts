import os from 'os'
import path from 'path'

const isDev: boolean = process.env.NODE_ENV === 'development'
const tmp: string = isDev ? path.resolve('.tmp') : os.tmpdir()
const home: string = path.join(os.homedir(), 'desktop')

const dataPaths = {
  design: path.resolve('__tests__', '_data', 'design.qr.svg'),
  designBarcode: path.resolve('__tests__', '_data', 'design.svg'),
  home,
  images: path.resolve('__tests__', '_data', 'images-qrcode'),
  imagesBarcode: path.resolve('__tests__', '_data', 'images-barcode'),
  key: path.resolve('__tests__', '_data', 'key.xlsx'),
  keyImage: path.resolve('__tests__', '_data', 'key.jpg'),
  questionsModel: isDev
    ? path.resolve('src/data/questions-model.json')
    : path.resolve('dist/data/questions-model.json'),
  result: path.resolve('__tests__', '_data', 'result.xlsx'),
  resultCompiled: path.resolve('__tests__', '_data', 'resultCompiled.xlsx'),
  root: path.resolve('.'),
  temp: tmp,
  tmp,
}

export { dataPaths }
