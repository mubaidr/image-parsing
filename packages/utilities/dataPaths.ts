import os from 'os';
import path from 'path';

const tmp = os.tmpdir();

export const DataPaths = {
  design: path.resolve('tests', '_data', 'design.qr.svg'),
  designBarcode: path.resolve('tests', '_data', 'design.svg'),
  home: os.homedir(),
  images: path.resolve('tests', '_data', 'images-qrcode'),
  imagesBarcode: path.resolve('tests', '_data', 'images-barcode'),
  key: path.resolve('tests', '_data', 'key.xlsx'),
  keyImage: path.resolve('tests', '_data', 'key.jpg'),
  result: path.resolve('tests', '_data', 'result.xlsx'),
  resultCompiled: path.resolve('tests', '_data', 'resultCompiled.xlsx'),
  root: path.resolve('.'),
  temp: tmp,
  tmp,
};
