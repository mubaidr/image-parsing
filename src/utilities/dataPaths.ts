import os from 'os'
import path from 'path'

const isDev: boolean = process.env.NODE_ENV === 'development'

const tmp: string = isDev ? path.resolve('.tmp') : os.tmpdir()
const home: string = path.join(os.homedir(), 'desktop')

const dataPaths = {
  design: path.resolve('_test_data', 'design.qr.svg'),
  designBarcode: path.resolve('_test_data', 'design.svg'),
  home,
  images: path.resolve('_test_data', 'images-qrcode'),
  imagesBarcode: path.resolve('_test_data', 'images-barcode'),
  key: path.resolve('_test_data', 'key.xlsx'),
  keyImage: path.resolve('_test_data', 'key.jpg'),
  questionsModel: path.resolve('src/data/questions-model.json'),
  result: path.resolve('_test_data', 'result.xlsx'),
  root: path.resolve('.'),
  temp: tmp,
  tmp,
}

export { dataPaths }
