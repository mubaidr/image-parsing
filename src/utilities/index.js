const brain = require('brain.js')
const cheerio = require('cheerio')
const childProcess = require('child_process')
const fastGlob = require('fast-glob')
const fs = require('fs')
const javascriptBarcodeReader = require('javascript-barcode-reader')
const os = require('os')
const download = require('downloadjs')
const NO_OF_CORES = require('physical-cpu-count')
const dataPaths = require('./data-paths')
// const javascriptBarcodeReader = require('../../../Javascript-Barcode-Reader/src')

/**
 * Create worker process equal to cpu cores
 *
 * @returns {Array} array of child process forks
 */
async function createWorkerProcesses(imagesCount) {
  const WORKERS = []

  const availMemory = (os.totalmem() - os.freemem()) / (1024 * 1024 * 1024)
  let CORE_COUNT = Math.min(NO_OF_CORES, imagesCount)

  // If available ram is less than 200MB/400MB, use only one/two worker processes respectively
  if (availMemory < 0.2) {
    CORE_COUNT = 1
  } else if (availMemory < 0.4) {
    CORE_COUNT = 2
  }

  for (let i = 0; i < CORE_COUNT; i += 1) {
    WORKERS.push(
      childProcess.fork(`${__dirname}/processTaskWorker.js`, [], {
        silent: true,
      }),
    )
  }

  return WORKERS
}

/**
 * Extracts position & dimensions of objects from SVG design File
 *
 * @param {String} designFilePath Path to the svg design file
 * @returns {Object} JSON object
 */
async function getDesignData(dir) {
  const designData = {
    questions: {},
  }
  const ROLL_NO_PATTERN = new RegExp(/rollnobarcode/gi)
  const QUESTION_PATTERN = new RegExp(/(q[1-9][0-9]?[ad])\b/gi)

  const $ = cheerio.load(fs.readFileSync(dir, 'utf8'))
  const svgViewBox = $('svg')[0].attribs.viewBox.split(' ')

  designData.width = Math.ceil(svgViewBox[2] - svgViewBox[0])
  designData.height = Math.ceil(svgViewBox[3] - svgViewBox[1])

  let gotRollNo = false
  let x
  let y
  let rx
  let ry
  let width
  let height
  let left
  let top

  const groups = $('g')
  for (let i = 0; i < groups.length; i += 1) {
    const group = groups[i]

    const title = $(group)
      .find('title')
      .first()
      .html()
      .trim()
      .toLowerCase()

    const isQuestionGroup = QUESTION_PATTERN.test(title)
    const isRollNoGroup =
      isQuestionGroup || gotRollNo ? false : ROLL_NO_PATTERN.test(title)

    if (isQuestionGroup || isRollNoGroup) {
      const transform = $(group)
        .attr('transform')
        .replace(/(translate)|\(|\)/gi, '')
        .split(',')
        .map(val => parseFloat(val))

      const rect = $(group)
        .find('rect')
        .first()

      left = parseFloat(rect.attr('x'))
      top = parseFloat(rect.attr('y'))

      rx = parseFloat(rect.attr('rx')) || 0
      ry = parseFloat(rect.attr('ry')) || 0

      x = left - rx + transform[0]
      y = top - ry + transform[1]

      width = parseFloat(rect.attr('width')) + rx
      height = parseFloat(rect.attr('height')) + ry
    }

    if (isQuestionGroup) {
      const optionTitle = title.slice(-1)
      const questionNumber = title.slice(0, -1)

      if (!designData.questions[questionNumber]) {
        designData.questions[questionNumber] = {}
      }

      if (optionTitle === 'a') {
        designData.questions[questionNumber].x1 = x
        designData.questions[questionNumber].y1 = y
      } else {
        designData.questions[questionNumber].x2 = x + width
        designData.questions[questionNumber].y2 = y + height
      }
    }

    if (isRollNoGroup) {
      designData.rollNo = { x1: x, y1: y, x2: x + width, y2: y + height }
      gotRollNo = true
    }
  }

  return designData
}

/**
 * Return a list of valid image format files from the provided path
 *
 * @param {String} path Path to sarch for images
 * @param {Array.<String>} format Array of extensions of valid image formats
 * @returns {Array.<String>} List of file paths
 */
function getImagePaths(dir) {
  const formats = [
    'png',
    'jpg',
    'jpeg',
    'jpe',
    // 'jfif',
    'gif',
    'tif',
    'tiff',
    'bmp',
    // 'dib',
  ]

  return fastGlob(`${dir}/*.{${formats.join(',')}}`, { onlyFiles: true })
}

/**
 * Returns a trained neural network function
 *
 * @returns {Function} Neural network function
 */
function getNeuralNet(src) {
  const net = new brain.NeuralNetwork()

  return net.fromJSON(
    JSON.parse(fs.readFileSync(src || dataPaths.trainingData)),
  )
}

/**
 *
 *
 * @param {Object} designData A JSON Object containing information about the position, width, height of elements in svg design file (available from utiltities/getDesignData)
 * @param {String} path Path of scanned image file
 * @param {Object=} resultsData Path to csv file for training data
 * @param {Number=} rollNo Roll no of the current scanned image
 * @returns {Object} {title: {String}, data: {buffer}}
 */
async function getQuestionsData(designData, img, resultsData, rollNo) {
  const SCALE = 0.25
  const IS_TEST_DATA = resultsData && rollNo

  return new Promise((resolveCol, rejectCol) => {
    img.resize(designData.width * SCALE).max()

    const promises = []
    // extract all questions portions
    Object.keys(designData.questions).forEach(title => {
      const p = new Promise(resolve => {
        const q = designData.questions[title]

        img
          .extract({
            left: Math.floor(q.x1 * SCALE),
            top: Math.floor(q.y1 * SCALE),
            width: Math.ceil((q.x2 - q.x1) * SCALE),
            height: Math.ceil((q.y2 - q.y1) * SCALE),
          })
          // .png()
          // .toFile(path.join(dataPaths.tmp, `${rollNo}-${title}.png`), err => {
          //   if (err) console.log(err)
          //   resolve()
          // })
          .toBuffer({ resolveWithObject: true })
          .then(({ data }) => {
            const binaryData = []

            for (let i = 0; i < data.length; i += 3) {
              const r = i
              const g = i + 1
              const b = i + 2
              const avg = Math.ceil((data[r] + data[g] + data[b]) / 3)
              const threshold = 15

              if (avg <= 80) {
                // black pixel
                binaryData.push(0)
              } else if (
                data[r] <= avg + threshold &&
                data[r] >= avg - threshold &&
                (data[g] <= avg + threshold && data[g] >= avg - threshold) &&
                (data[b] <= avg + threshold && data[b] >= avg - threshold)
              ) {
                // grey pixel
                binaryData.push(1)
              } else {
                // color pixel
                binaryData.push(0)
              }
            }

            if (IS_TEST_DATA) {
              // for training data
              if (resultsData[rollNo] && resultsData[rollNo][title] !== '*') {
                const o = {}
                o[resultsData[rollNo][title]] = 1

                resolve({ input: binaryData, output: o })
              } else {
                resolve(false)
              }
            } else {
              // for processing data
              resolve({ title, data: binaryData })
            }
          })
      })

      promises.push(p)
    })

    Promise.all(promises)
      .then(res => {
        resolveCol(res)
      })
      .catch(err => {
        rejectCol(err)
      })
  })
}

/**
 *
 *
 * @param {Object} designData A JSON Object containing information about the position, width, height of elements in svg design file (available from utiltities/getDesignData)
 * @param {String} path Path of scanned image file
 * @returns {String} Roll Number
 */
async function getRollNoFromImage(designData, img) {
  const metadata = await img.metadata()
  const rollNoPos = designData.rollNo
  const ratio = metadata.width / designData.width
  const width = Math.ceil((rollNoPos.x2 - rollNoPos.x1) * ratio)
  const height = Math.ceil((rollNoPos.y2 - rollNoPos.y1) * ratio)

  const { data, info } = await img
    .extract({
      left: Math.floor(rollNoPos.x1 * ratio),
      top: Math.floor(rollNoPos.y1 * ratio),
      width,
      height,
    })
    .toBuffer({ resolveWithObject: true })

  const newData = []
  if (info.channels < 4) {
    for (let i = 0; i < data.length; i += info.channels) {
      // copy current channels
      for (let j = 0; j < info.channels; j += 1) {
        newData.push(data[i + j])
      }
      // add missing channels
      for (let k = 0; k < 4 - info.channels; k += 1) {
        newData.push(255)
      }
    }
  }

  return javascriptBarcodeReader(
    { data: info.channels < 4 ? newData : data, width, height },
    { barcode: 'code-39' },
  )
}

/**
 *
 *
 * @param {String} str JSON object or JSON file path
 * @param {Boolean} isPath True for path
 * @param {Boolean} isKey True if the CSV file is answer key file
 * @returns {Object} CSV String
 */
async function CSVToJSON(str, isPath, isKey) {
  const resultFile = isPath ? fs.readFileSync(str, 'utf8') : str
  const json = {}

  // extract rows and header names
  const rows = resultFile.split('\n')
  const headerValues = rows[0]
    .replace(/ /gi, '')
    .toLowerCase()
    .split(',')

  // extract index of roll no in header row
  const rollNoIndex = Math.max(
    headerValues.indexOf('rollno'),
    headerValues.indexOf('roll#'),
    headerValues.indexOf('rollnumber'),
  )

  if (isKey) rows.length = 2
  // iterate rows but skips header row
  for (let i = 1; i < rows.length; i += 1) {
    const obj = {}
    const values = rows[i]
      .replace(/(\s)|(\.)|(-)|(_)/gi, '')
      // .toLowerCase()
      .split(',')

    values.forEach((value, index) => {
      if (index !== rollNoIndex) {
        obj[headerValues[index]] = value
      }
    })

    if (isKey) {
      json.key = obj
    } else {
      const roll = values[rollNoIndex]

      if (roll) {
        json[roll] = obj
      }
    }
  }

  return json
}

/**
 *
 *
 * @param {String} str JSON object or JSON file path
 * @param {Boolean} isPath True for path
 * @returns {Object} CSV String
 */
function JSONToCSV(str, isPath) {
  const obj = isPath ? JSON.parse(fs.readFileSync(str, 'utf8')) : str

  let header = 'ROLL NO'
  let csv = ''

  // sort by roll number
  obj.sort((a, b) => Object.keys(a)[0] - Object.keys(b)[0])

  // parse to csv
  Object.values(obj).forEach((val, index) => {
    const [[rollno, keys]] = Object.entries(val)

    // eslint-disable-next-line
    csv += '\n'
    csv += rollno

    // sort by question no
    const keyEntries = Object.entries(keys).sort(
      (a, b) => a[0].replace('q', '') - b[0].replace('q', ''),
    )

    // append to csv
    for (let i = 0; i < keyEntries.length; i += 1) {
      if (i !== keyEntries.length) {
        csv += ','
      }

      csv += keyEntries[i][1].toUpperCase
        ? keyEntries[i][1].toUpperCase()
        : keyEntries[i][1]

      if (index === 0) {
        header += `,${keyEntries[i][0].toUpperCase()}`
      }
    }
  })

  return `${header}${csv}`
}

/**
 * Exports result data
 *
 * @param {Object} resultData Result data JSON format
 * @param {String} name Name of the output file
 * @param {Boolean} isCSV if the provided data is CSV
 */
async function exportResult(resultData, name, isCSV) {
  const csv = isCSV ? resultData : await JSONToCSV(resultData)

  try {
    download(csv, name ? `${name}.csv` : 'output.csv', 'text/csv')
  } catch (error) {
    console.log('output: ', csv)
  }
}

/**
 * Compare the provided result with key to generate marks
 *
 * @param {Object} keys JSON Object containing answer keys
 * @param {Object} result JSON Object containing result keys
 * @param {Number} correctMarks Number of marks to assign per correct answer
 * @param {Number} negativeMarks Number of marks to deduct per correct answer
 * @returns {Number} Total marks obtained
 */
async function compileResult(keys, results, correctMarks, negativeMarks) {
  const keyEntries = Object.entries(keys.key)
  const output = []
  let totalMarks = 0

  for (let i = 0; i < keyEntries.length; i += 1) {
    const [, value] = keyEntries[i]

    if (value === '?' || value === '*') continue

    totalMarks += correctMarks
  }

  Object.entries(results).forEach(([rollNo, answers]) => {
    let marks = 0

    for (let i = 0; i < keyEntries.length; i += 1) {
      const [key, value] = keyEntries[i]

      // if user has not selected any option
      if (value === '?') continue
      // if question has multiple correct options
      if (value === '*') continue
      // if question has no right option
      if (answers[key] === '?') continue

      if (answers[key] === value) {
        // correct option
        marks += correctMarks
      } else {
        // in-correct option
        marks -= negativeMarks
      }
    }

    output.push({
      [rollNo]: {
        marks,
        totalMarks,
      },
    })
  })

  // export
  exportResult(await JSONToCSV(output), 'ResultMarks', true)
}

module.exports = {
  createWorkerProcesses,
  getDesignData,
  getImagePaths,
  getNeuralNet,
  getRollNoFromImage,
  CSVToJSON,
  JSONToCSV,
  getQuestionsData,
  exportResult,
  compileResult,
}
