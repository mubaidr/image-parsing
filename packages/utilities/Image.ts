import fastGlob from 'fast-glob';
import path from 'path';
import type { Sharp } from 'sharp';
import sharp from 'sharp';
import { v4 as uuid4 } from 'uuid';
import { DataPaths } from './dataPaths';

export enum ImageNativeTypes {
  bmp,
  gif,
  jfif,
  jpe,
  jpeg,
  jpg,
  png,
  svg,
  webp,
}

export enum ImageTypes {
  bmp,
  dib,
  gif,
  jfif,
  jpe,
  jpeg,
  jpg,
  png,
  svg,
  tif,
  tiff,
  webp,
}

// average using linear approximation of gamma & perceptual luminance correction
function pixelAverage(r: number, g: number, b: number): number {
  return Math.floor(0.299 * r + 0.587 * g + 0.114 * b);
}

export class Image {
  static TARGET_SIZE = 1280;
  static CHANNELS = 3;

  public id: string;
  public source: string;
  public isNative = false;
  public channels = Image.CHANNELS;
  public width = 0;
  public height = 0;
  public data: Uint8ClampedArray = Uint8ClampedArray.from([]);

  constructor(source: string) {
    this.id = uuid4();
    this.source = source;

    const extension = source.split('.').pop();
    if (extension && extension in ImageNativeTypes) {
      this.isNative = true;
    }
  }

  static readDirectory(dir: string): string[] {
    const loc = dir.replace(/\\/g, '/');
    const exts = Object.keys(ImageTypes).map((ext) => `.${ext}`);
    const glob = `${loc}/**/*{${exts}}`.replace('//', '/');

    return fastGlob.sync(glob, {
      absolute: true,
      onlyFiles: true,
      stats: false,
    });
  }

  static async load(source: string): Promise<Image> {
    const { data, info } = await sharp(source)
      .resize(Image.TARGET_SIZE, null, {
        kernel: sharp.kernel.mitchell,
        fastShrinkOnLoad: false,
        withoutEnlargement: true,
      })
      .flatten()
      .median(3)
      .raw()
      .toBuffer({ resolveWithObject: true });

    const image = new Image(source);

    image.data = Uint8ClampedArray.from(data);
    image.width = info.width;
    image.height = info.height;
    image.channels = info.channels;

    return image;
  }

  async log(name: string = uuid4(), location?: string): Promise<string> {
    const target = path.join(location || DataPaths.tmp, `${name}.jpg`);

    await this.toSharpObject()
      .jpeg()
      .toFile(target)
      .then(() => {
        console.info(`[log] : ${target}`);
      })
      .catch(console.error);

    return target;
  }

  toSharpObject(): Sharp {
    return sharp(Buffer.from(this.data), {
      raw: {
        width: this.width,
        height: this.height,
        channels: this.channels as 1 | 2 | 3 | 4,
      },
    });
  }

  clone(data?: Uint8ClampedArray, width?: number, height?: number): Image {
    const image = new Image(this.source);

    image.isNative = this.isNative;
    image.width = width ? width : this.width;
    image.height = height ? height : this.height;
    image.data = data ? data : Uint8ClampedArray.from([...this.data]);

    return image;
  }

  greyscale(): Image {
    return this.grayscale();
  }

  grayscale(): Image {
    for (let i = 0; i < this.data.length; i += this.channels) {
      const [r, g, b] = this.data.slice(i, i + this.channels);
      this.data[i] =
        this.data[i + 1] =
        this.data[i + 2] =
          pixelAverage(r, g, b);
    }

    return this;
  }

  toBitImage(): Image {
    const data: number[] = [];

    for (let i = 0; i < this.data.length; i += this.channels) {
      const [r, g, b] = this.data.slice(i, i + this.channels);
      data.push(pixelAverage(r, g, b));
    }

    this.channels = 1;
    this.data = new Uint8ClampedArray(data);

    return this;
  }

  getPercentFilled(): number {
    let sum = 0;

    for (let i = 0; i < this.data.length; i += this.channels) {
      const [r, g, b] = this.data.slice(i, i + this.channels);
      const avg = pixelAverage(r, g, b);
      const threshold = 25;
      const thresholdBlack = 75;
      const upperLimit = avg + threshold;
      const lowerLimit = avg - threshold;

      if (avg <= thresholdBlack) {
        // Black pixel
        sum += 1;
      } else if (
        r <= upperLimit &&
        r >= lowerLimit &&
        g <= upperLimit &&
        g >= lowerLimit &&
        b <= upperLimit &&
        b >= lowerLimit
      ) {
        // Grey pixel
        continue;
      } else {
        // Color pixel
        sum += 1;
      }
    }

    return (sum / (this.data.length / 3)) * 100;
  }

  extract(x = 0, y = 0, width = this.width, height = this.height): Image {
    const data = new Uint8ClampedArray(width * height * this.channels);

    for (let top = y; top < y + height; top += 1) {
      const start = (top * this.width + x) * this.channels;
      const end = (top * this.width + x + width - 1) * this.channels;
      const row = top - y;
      data.set(this.data.slice(start, end), row * width * this.channels);
    }

    return this.clone(data, width, height);
  }

  // getShapes<T extends ShapeTypes>(shapeType: ShapeTypes): T[] {
  //   return contoursUtility.getShapes<T>(this.clone().grayscale(), shapeType)
  // }
}
