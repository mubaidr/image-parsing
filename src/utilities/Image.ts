import fastGlob from 'fast-glob'
// import NodeCache from 'node-cache'
import path from 'path'
import sharp from 'sharp'
import { v4 as uuid4 } from 'uuid'
import { DataPaths } from './dataPaths'

// const myCache = new NodeCache()

export enum ImageNativeTypes {
  bmp,
  gif,
  jfif,
  jpe,
  jpeg,
  jpg,
  png,
  svg,
  webp,
}

export enum ImageTypes {
  bmp,
  dib,
  gif,
  jfif,
  jpe,
  jpeg,
  jpg,
  png,
  svg,
  tif,
  tiff,
  webp,
}

export class Image {
  static TARGET_SIZE = 1240
  id: string
  source: string
  isNative = false
  width = 0
  height = 0
  data: Uint8ClampedArray = Uint8ClampedArray.from([])

  constructor(source: string) {
    this.id = uuid4()
    this.source = source

    const extension = source.split('.').pop()
    if (extension && extension in ImageNativeTypes) {
      this.isNative = true
    }
  }

  static readDirectory(dir: string): string[] {
    const loc = dir.replace(/\\/g, '/')
    const exts = Object.keys(ImageTypes).map((ext) => `.${ext}`)
    const glob = `${loc}/**/*{${exts}}`.replace('//', '/')

    return fastGlob.sync(glob, {
      absolute: true,
      onlyFiles: true,
      stats: false,
    })
  }

  static async load(source: string, avoidResize = false): Promise<Image> {
    const sharpImage = sharp(source)

    if (!avoidResize) sharpImage.resize(Image.TARGET_SIZE)

    const { data, info } = await sharpImage
      .raw()
      .flatten()
      .toBuffer({ resolveWithObject: true })

    const image = new Image(source)

    image.data = Uint8ClampedArray.from(data)
    image.width = info.width
    image.height = info.height

    return image
  }

  log(name: string = uuid4(), location?: string): void {
    const target = path.join(location || DataPaths.tmp, `${name}.jpg`)
    // BUG: Sharp instance
    // await sharp(src).jpeg().toFile(target)
    console.log(`[log] : ${target}`)
  }

  clone(data?: Uint8ClampedArray, width?: number, height?: number): Image {
    const image = new Image(this.source)
    image.isNative = this.isNative
    image.width = width ? width : this.width
    image.height = height ? height : this.height
    image.data = data ? data : Uint8ClampedArray.from([...this.data])

    return image
  }

  grayscale(): Image {
    for (let i = 0; i < this.data.length; i += 3) {
      const brightness =
        0.34 * this.data[i] + 0.5 * this.data[i + 1] + 0.16 * this.data[i + 2]

      this.data[i] = this.data[i + 1] = this.data[i + 2] = brightness
    }

    return this
  }

  getPercentFilled(): number {
    let sum = 0

    for (let i = 0; i < this.data.length; i += 3) {
      const [r, g, b] = this.data.slice(i, i + 3)
      const threshold = 15
      const thresholdBlack = 80
      const avg = Math.ceil(0.34 * r + 0.5 * g + 0.16 * b)
      const upperLimit = avg + threshold
      const lowerLimit = avg - threshold

      if (avg <= thresholdBlack) {
        // Black pixel
        sum += 1
      } else if (
        r <= upperLimit &&
        r >= lowerLimit &&
        g <= upperLimit &&
        g >= lowerLimit &&
        b <= upperLimit &&
        b >= lowerLimit
      ) {
        // Grey pixel
        continue
      } else {
        // Color pixel
        sum += 1
      }
    }

    return (sum / (this.data.length / 3)) * 100
  }

  extract(x = 0, y = 0, width = this.width, height = this.height): Image {
    const data = new Uint8ClampedArray(width * height * 3)

    for (let top = y; top < y + height; top += 1) {
      const start = (top * this.width + x) * 3
      const end = (top * this.width + x + width - 1) * 3
      const row = top - y
      data.set(this.data.slice(start, end), row * width * 3)
    }

    return this.clone(data, width, height)
  }
}
