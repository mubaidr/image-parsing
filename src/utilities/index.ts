import brain, { INeuralNetworkJSON } from 'brain.js'
import fs from 'fs'

import Result from './@classes/Result'
import KeyNativeEnum from './@enums/KeyNativeEnum'
import { dataPaths } from './dataPaths'
import { getDesignData } from './design'
import { importExcelToJson } from './excel'
import extractTask from './workers/workerExtract'

const getQuestionsNeuralNet = (): brain.NeuralNetwork => {
  const text = fs.readFileSync(dataPaths.questionsModel).toString()
  const json: INeuralNetworkJSON = JSON.parse(text)

  return new brain.NeuralNetwork().fromJSON(json)
}

const convertToBitArray = (data: number[], channels: number): number[] => {
  const binaryData: number[] = []

  for (let i = 0, dataLength = data.length; i < dataLength; i += channels) {
    const threshold = 15
    const thresholdBlack = 80
    const [r, g, b] = data.slice(i, i + channels)
    const avg = Math.ceil((r + g + b) / channels)
    const upperLimit = avg + threshold
    const lowerLimit = avg - threshold

    if (avg <= thresholdBlack) {
      // Black pixel
      binaryData.push(0)
    } else if (
      r <= upperLimit &&
      r >= lowerLimit &&
      g <= upperLimit &&
      g >= lowerLimit &&
      b <= upperLimit &&
      b >= lowerLimit
    ) {
      // Grey pixel
      binaryData.push(1)
    } else {
      // Color pixel
      binaryData.push(0)
    }
  }

  return binaryData
}

const readKey = async (src: string): Promise<Result[] | undefined> => {
  const ext = src.split('.').pop()

  if (ext === undefined) throw new Error('Invalid path specified')

  if (ext in KeyNativeEnum) {
    const rows = importExcelToJson(src)
    const results: Result[] = []

    for (let i = 0, len = rows.length; i < len; i += 1) {
      results.push(Result.fromJson(rows[i]))
    }

    return results
  }

  console.log('yehn group is here.... !!!!')

  const designData = getDesignData(dataPaths.designBarcode)
  const keys = await extractTask.start(
    {
      designData,
      imagePaths: [src],
    },
    false,
  )

  return keys
}

export { convertToBitArray, getQuestionsNeuralNet, readKey }
