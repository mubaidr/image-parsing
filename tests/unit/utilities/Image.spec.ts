import { DataPaths } from '@/utilities/dataPaths'
import { Image } from '@/utilities/Image'
import { existsSync } from 'fs'
import * as path from 'path'

describe('Image', () => {
  test('load', async () => {
    const image = await Image.load(DataPaths.keyImage)

    expect(image.width).toBe(Image.TARGET_SIZE)
    expect(image.source).toBe(DataPaths.keyImage)
    expect(image.isNative).toBeTruthy()
    expect(image.data.length * image.width * image.height).toBeGreaterThan(0)
  })

  test('grayscale', async () => {
    const image = await Image.load(DataPaths.keyImage)

    const { width, height } = image

    image.grayscale()

    expect(image.width * image.height).toBe(width * height)
    expect(image.data.length * image.width * image.height).toBeGreaterThan(0)
  })

  test('extract', async () => {
    const image = await Image.load(
      path.resolve('./tests/_data/empty-20x20.png')
    )

    const width = 10
    const height = 10

    // 5, 10, 10, 20
    const imageCopy = image.extract(width / 2, height / 2, width, height)

    imageCopy.log()

    expect(imageCopy.width * imageCopy.height).toBe(width * height)
    expect(imageCopy.data.length).toBe(width * height * 3)
    expect(imageCopy.data).toMatchSnapshot()
  })

  test('log', async () => {
    const image = await Image.load(
      path.resolve('./tests/_data/empty-20x20.png'),
      true
    )

    const width = 10
    const height = 10

    const imageCopy = image.extract(width / 2, height / 2, width, height)
    const target = await imageCopy.log()

    expect(existsSync(target)).toBeTruthy()
  })
})
