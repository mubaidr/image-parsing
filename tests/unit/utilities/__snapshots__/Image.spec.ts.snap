// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image extract 1`] = `
Uint8ClampedArray [
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
]
`;
